arch/arm/core/aarch64/smp.c

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2020 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+/**
+ * @file
+ * @brief codes required for AArch64 multicore and Zephyr smp support
+ */
+
+#include <cache.h>
+#include <device.h>
+#include <kernel.h>
+#include <kernel_structs.h>
+#include <ksched.h>
+#include <soc.h>
+#include <init.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <drivers/interrupt_controller/gic.h>
+#include <drivers/psci.h>
+#include <sys/arch_interface.h>
+
+volatile struct {
+	arch_cpustart_t fn;
+	void *arg;
+} arm64_cpu_init[CONFIG_MP_NUM_CPUS];
+
+volatile char *arm64_cpu_sp;
+
+volatile uint32_t arm64_cpu_up_flag;
+
+/*
+ * _curr_cpu is used to record the struct of _cpu_t of each cpu.
+ * for efficient usage in assembly
+ */
+volatile _cpu_t *_curr_cpu[CONFIG_MP_NUM_CPUS];
+
+extern void __start(void);
+/* Called from Zephyr initialization */
+void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
+		    arch_cpustart_t fn, void *arg)
+{
+	const struct device *psci;
+
+	_curr_cpu[cpu_num] = &(_kernel.cpus[cpu_num]);
+	arm64_cpu_init[cpu_num].fn = fn;
+	arm64_cpu_init[cpu_num].arg = arg;
+
+	arm64_cpu_sp = Z_THREAD_STACK_BUFFER(stack) + sz;
```

Comment:
I would mildly prefer here if `sp` was a member of `arm64_cpu_init`.

As-is, you're using `arm64_cpu_sp` as a global mailbox but not `arm64_cpu_init`, which is a little odd.

Then you could also, assuming you set up the ordering correctly, simplify this somewhat and just do

```c
arm64_cpu_init[cpu_num].fn = fn;
arm64_cpu_init[cpu_num].arg = arg;
arch_dcache_range(&arm64_cpu_init[cpu_num], sizeof(arm64_cpu_init[cpu_num]), K_CACHE_WB); // Assuming this is a memory barrier too. Famous last words.
arm64_cpu_init[cpu_num].sp = Z_THREAD_STACK_BUFFER(stack) + sz;
arch_dcache_range(&arm64_cpu_init[cpu_num].sp, sizeof(arm64_cpu_init[cpu_num].sp), K_CACHE_WB);

psci = device_get_binding("PSCI");

if (psci_cpu_on(psci, cpu_num, (uint64_t)&__start)) {
    printk("Failed to boot CPU%d\n", cpu_num);
}

// Probably want a timeout here
while (arm64_cpu_init[cpu_num].fn) {
    // probably unnecessary, but this is early boot for the other core so memory coherency is not a given
    arch_dcache_range(&arm64_cpu_init[cpu_num].fn, sizeof(arm64_cpu_init[cpu_num].fn), K_CACHE_WB_INVD);
}
```

where the secondary core portion of the boot sequence spins until `arm64_cpu_init[cpu_num].sp` is non-null (likely with a repeated cache line clean and invalidate + dmb in there, because again this is early boot and not always memory-coherent yet), then later on reads the value of `arm64_cpu_init[cpu_num]`, clears it, and does the actual `fn(arg)` call.

This avoids having to have separate variables for `arm64_cpu_sp` and `arm64_cpu_up_flag`, as well as laying the framework for having multiple CPUs concurrently start up (as now all of the variables here are per-core.).

Reply:
yes, I really like this solution. @MrVan any reason why you didn't consider this?